pepper-apis/src/main/java/org/broadinstitute/ddp/route/UserRegistrationRoute.java

```diff
@@ -225,6 +166,95 @@ public Object handle(Request request, Response response, UserRegistrationPayload
+        UserDao userDao = handle.attach(UserDao.class);
+        String invitationGuid = invitation.getInvitationGuid();
+        String studyGuid = study.getGuid();
+
+        if (invitation.getInvitationType() == InvitationType.AGE_UP) {
+            var user = userDao.findUserById(invitation.getUserId())
+                    .orElseThrow(() -> new DDPException("Could not find user " + invitation.getUserId()));
+            if (user.hasAuth0Account()) {
+                throw new DDPException("There is already an account-bearing user for invitation " + invitationGuid);
+            }
+
+            // verify that there is governance policy configured for the study and that the
+            // user has reached age of majority.
+            GovernancePolicy policy = handle.attach(StudyGovernanceDao.class).findPolicyByStudyGuid(studyGuid).orElse(null);
+            if (policy != null) {
+                if (policy.hasReachedAgeOfMajority(handle, interpreter, user.getGuid())) {
+                    LOG.info("Assigning {} to user {} for invitation {}", auth0UserId, user.getGuid(), invitationGuid);
+
+                    var numRows = userDao.updateAuth0UserId(user.getGuid(), auth0UserId);
+                    if (numRows != 1) {
+                        throw new DDPException("Updated " + numRows + " for " + auth0UserId);
+                    }
+                    LOG.info("User {} has been associated with auth0 id {}", user.getGuid(), auth0UserId);
+
+                    numRows = handle.attach(InvitationDao.class).updateAcceptedAt(invitation.getInvitationId(), Instant.now());
+                    if (numRows != 1) {
+                        throw new DDPException("Updated " + numRows + " for invitation " + invitationGuid);
+                    }
+
+                    EventSignal signal = new EventSignal(user.getId(), user.getId(), user.getGuid(),
+                            study.getId(), EventTriggerType.GOVERNED_USER_REGISTERED);
+                    EventService.getInstance().processAllActionsForEventSignal(handle, signal);
+                } else {
+                    LOG.error("User {} is not allowed to create an account yet because they have not reached age of majority "
+                            + " in study {} with invitation {}", user.getGuid(), studyGuid, invitationGuid);
+                    ResponseUtil.halt422ErrorResponse(response, ErrorCodes.GOVERNANCE_POLICY_VIOLATION);
+                }
+            } else {
+                LOG.error("No governance policy for study {}.  Why is a client registering user {} with invitation {} ?",
+                        studyGuid, auth0UserId, invitationGuid);
+                ResponseUtil.halt422ErrorResponse(response, ErrorCodes.GOVERNANCE_POLICY_VIOLATION);
+            }
+
+            return user;
+        } else if (invitation.getInvitationType() == InvitationType.RECRUITMENT) {
```

Comment:
I like the introduction of a new type.  Maybe introduce an interface like `InvitationValidator` instead of a type-based if-then?

Reply:
We have already done some sanity checks and validation, so this part of the code is for assigning the invitation.

Reply:
Is there something we'll like changed here?